### PR TITLE
contracts: Allow stack height metering to be disabled

### PR DESCRIPTION
**Pull request by athei**
Stack height metering is only needed to combat indeterminism of different execution engine versions. Since we put wasmi inside the runtime for the time being we don't it anymore.

This PR allows stack height metering to be disabled by the runtime configuration and also disables it by default. This reduces the cached contract size by 20% to 40%.

### CODE REVIEW

File: frame/contracts/src/weights.rs [hunk 1]
```diff
@@ -160,32 +160,32 @@ pub struct SubstrateWeight<T>(PhantomData<T>);
...
 			.saturating_add(T::DbWeight::get().writes(1 as Weight))
 	}
 	// Storage: Contracts PristineCode (r:1 w:0)
 	// Storage: Contracts CodeStorage (r:0 w:1)
 	fn reinstrument(c: u32, ) -> Weight {
-		(17_839_000 as Weight)
+		(7_105_000 as Weight)
 			// Standard Error: 0
-			.saturating_add((64_000 as Weight).saturating_mul(c as Weight))
+			.saturating_add((50_000 as Weight).saturating_mul(c as Weight))
```
Comment (athei) [inline frame/contracts/src/weights.rs:188]: Expected: Instrumentation gets cheaper.

File: frame/contracts/src/weights.rs [hunk 2]
```diff
@@ -234,25 +234,25 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	// Storage: Timestamp Now (r:1 w:0)
 	// Storage: System Account (r:1 w:1)
 	fn call() -> Weight {
-		(140_234_000 as Weight)
+		(189_970_000 as Weight)
 			.saturating_add(T::DbWeight::get().reads(4 as Weight))
 			.saturating_add(T::DbWeight::get().writes(2 as Weight))
 	}
```
Comment (athei) [inline frame/contracts/src/weights.rs:240]: I don't get it. The weight for a contract call goes up. This should be unrelated to this change. Same across other benchmarking results. There is some static cost added to a contract call. Maybe something merged in the meantime without a benchmark?

File: frame/contracts/src/weights.rs [hunk 3]
```diff
@@ -234,25 +234,25 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
...
 	// Storage: Contracts OwnerInfoOf (r:0 w:1)
 	fn upload_code(c: u32, ) -> Weight {
-		(49_467_000 as Weight)
+		(46_039_000 as Weight)
 			// Standard Error: 0
-			.saturating_add((66_000 as Weight).saturating_mul(c as Weight))
+			.saturating_add((50_000 as Weight).saturating_mul(c as Weight))
 			.saturating_add(T::DbWeight::get().reads(1 as Weight))
 			.saturating_add(T::DbWeight::get().writes(3 as Weight))
 	}
```
Comment (athei) [inline frame/contracts/src/weights.rs:250]: This is expected. The instrumentation cost goes down.

File: frame/contracts/src/weights.rs [hunk 4]
```diff
@@ -208,9 +208,9 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
...
 	// Storage: Contracts OwnerInfoOf (r:0 w:1)
 	fn instantiate_with_code(c: u32, s: u32, ) -> Weight {
-		(283_144_000 as Weight)
+		(379_283_000 as Weight)
 			// Standard Error: 0
-			.saturating_add((140_000 as Weight).saturating_mul(c as Weight))
+			.saturating_add((128_000 as Weight).saturating_mul(c as Weight))
 			// Standard Error: 0
 			.saturating_add((2_000 as Weight).saturating_mul(s as Weight))
 			.saturating_add(T::DbWeight::get().reads(5 as Weight))
```
Comment (athei) [inline frame/contracts/src/weights.rs:216]: This is expected. The instrumentation cost goes down. However, the overall cost goes up. No idea where this comes from.

**Comment by athei** [comment]
I did run the benchnarks against a unmodified master and they looked normal: https://github.com/paritytech/substrate/pull/10881

I then ran the bot here again and they look like I expected them: Costs for instrumentation goes down the rest stays the same. I figured there was something going on with the benchmarking machine.

File: frame/contracts/src/weights.rs [hunk 5]
```diff
@@ -160,32 +160,32 @@ pub struct SubstrateWeight<T>(PhantomData<T>);
...
 	// Storage: Contracts CodeStorage (r:0 w:1)
 	fn reinstrument(c: u32, ) -> Weight {
-		(17_839_000 as Weight)
+		(15_212_000 as Weight)
 			// Standard Error: 0
-			.saturating_add((64_000 as Weight).saturating_mul(c as Weight))
+			.saturating_add((51_000 as Weight).saturating_mul(c as Weight))
 			.saturating_add(T::DbWeight::get().reads(1 as Weight))
 			.saturating_add(T::DbWeight::get().writes(1 as Weight))
 	}
```
Comment (athei) [inline frame/contracts/src/weights.rs:191]: Expected: Instrumentation weight per byte goes down by 30%.

File: frame/contracts/src/weights.rs [hunk 6]
```diff
@@ -234,25 +234,25 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
...
 	// Storage: Contracts CodeStorage (r:1 w:1)
 	// Storage: Contracts PristineCode (r:0 w:1)
 	// Storage: Contracts OwnerInfoOf (r:0 w:1)
 	fn upload_code(c: u32, ) -> Weight {
-		(49_467_000 as Weight)
+		(44_290_000 as Weight)
 			// Standard Error: 0
-			.saturating_add((66_000 as Weight).saturating_mul(c as Weight))
+			.saturating_add((51_000 as Weight).saturating_mul(c as Weight))
```
Comment (athei) [inline frame/contracts/src/weights.rs:247]: Some improvement can be seen when uploading the code.

**Review by ascjones — APPROVED** [review]

**Comment by athei** [comment]
bot merge

**Pull request merged**